### PR TITLE
feat: add documentation for SplitChunksPlugin and deprecation warning…

### PR DESCRIPTION
**Pull request by jeremenichelli**
Sorry for the noise, the content for this branch was already approved here: https://github.com/webpack/webpack.js.org/pull/1863 but because of some issues on rebasing I needed to create a new PR to make sure I'm not making regression on content after the `next` branch merge.

Solves https://github.com/webpack/webpack.js.org/issues/1812

### CODE REVIEW

**Comment by jeremenichelli** [comment]
Hey @skipjack, Travis is throwing an error: `API rate limit exceeded for 35.224.112.202`, did we exceeded some quota around the service? 🤔

**Comment by montogeek** [comment]
@jeremenichelli It is exceeded every a few hours, something to fix in Infrastructure

**Comment by jeremenichelli** [comment]
@montogeek should we just trigger a new build for the branch?

**Comment by montogeek** [comment]
@jeremenichelli I am fixing it for once and all https://github.com/webpack/webpack.js.org/pull/1963

File: src/content/plugins/split-chunks-plugin.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+title: SplitChunksPlugin
+contributors:
+  - sokra
+  - jeremenichelli
+related:
+  - title: "webpack 4: Code Splitting, chunk graph and the splitChunks optimization"
+    url: https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366
+---
+
+Originally in webpack, chunks (and modules imported inside them) were connected by a parent-child relationship in the internal webpack graph.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: I would rephrase as (merged with the next lines as well):

> By default, chunks (and the modules inside them) are connected by direct parent-child relationships in the dependency graph. However, there are instances where a dependency may be duplicated across chunks. Previously, the `CommonsChunkPlugin` was use to dedupe these dependencies. In v4, the `optimization.splitChunks` and `optimization.runtimeChunk` configuration options provide a more intuitive interface.
>
> Here is how the new flow works...

File: src/content/plugins/split-chunks-plugin.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+This connection didn't allow to further optimizations and resulted into more downloaded code.
+
+webpack v4 removes the `CommonsChunkPlugin` in favor of `optimization.splitChunks` and `optimize.runtimeChunk` options. Here is how the new flow works.
+
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: We typically use two line breaks between `h2` sections. We should really automate this via `remark` and/or `prettier` along with various other markdown guidelines.

File: src/content/plugins/split-chunks-plugin.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+Why:
+
+* Condition 1: The chunk is shared between both import calls
+* Condition 2: `helpers` is bigger than 30kb
+* Condition 3: Number of parallel requests at the import calls is 2
+* Condition 4: Doesn't affect request at initial page load
+
+Putting the content of `helpers` into each chunk will result into its code being downloaded twice. By using a separate chunk this will only happen once. We pay the cost of an additional request, which could be considered a tradeoff. That's why there is a minimum size of 30kb.
+
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: Same two line breaks nit (just helps with readability when editing).

File: src/content/plugins/split-chunks-plugin.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+Originally in webpack, chunks (and modules imported inside them) were connected by a parent-child relationship in the internal webpack graph.
+
+This connection didn't allow to further optimizations and resulted into more downloaded code.
+
+webpack v4 removes the `CommonsChunkPlugin` in favor of `optimization.splitChunks` and `optimize.runtimeChunk` options. Here is how the new flow works.
+
+## Defaults
+
+Out of the box `SplitChunksPlugin` should work great for most users.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: Is it included out of the box or only when certain flags/config are present? Or do you just mean when included with no options, like so:

``` js
plugins: [
  new SplitChunksPlugin()
]
```

File: src/content/plugins/split-chunks-plugin.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+By default it only affects on-demand chunks because changing initial chunks would affect the script tags the HTML file should include to run the project.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: Maybe explain why changing those initial script tags is a problem?

File: src/content/plugins/split-chunks-plugin.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+		commons: {
+			test: /[\\/]node_modules[\\/],
+			name: "vendors",
+			chunks: "all"
+		}
+	}
+}
+```
+
+W> This downloads more code than neccessary.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: s/neccessary/necessary

Comment (skipjack) [reply src/content/plugins/split-chunks-plugin.md]: Also maybe elaborate this warning a bit...

> This will probably yield an unnecessarily large bundle as many applications use a bunch of `node_modules`. It's typically better to extract and load your core frameworks on initial load, and let other dependencies be chunked out and loaded dynamically later on.

File: src/content/plugins/split-chunks-plugin.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+You can set `optimization.splitChunks.cacheGroups.default` to `false` to disable the default cache group, same for `vendors` cache group.
+
+The priority of the default groups are negative to allow any custom cache group to take higher priority (the default value is `0`).
+
+Here are some examples and their effect:
+
+### Split Chunks: Example 1
+
+Create a `commons` chunk, which includes all code shared between entrypoints.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: s/entrypoints/entry points

**Comment by phun-ky** [comment]
Nice work, this is very much needed for all webpack4 users!

But, I can't see that you have included all the options for splitchunkcs, for example:

`automaticNameDelimiter` (defaults to: ~)

All options is in: `schemas\WebpackOptions.json`, and the default options are set here: `lib\WebpackOptionsDefaulter.js`

Specially `automaticNameDelimiter` was important to us, because we have a WAF (NAXSI) that had a ruled triggered for the usage of the tilde(~) character, resulting in a 406 Not Acceptable on the server when the chunk is requested: `vendor~moduleA~moduleB.js`.

File: src/content/plugins/commons-chunk-plugin.md [hunk 8]
```diff
@@ -6,9 +6,14 @@ contributors:
   - christopher4lis
   - kevinzwhuang
   - jdbevan
+  - jeremenichelli
 ---
 
-The `CommonsChunkPlugin` is an opt-in feature that creates a separate file (known as a chunk), consisting of common modules shared between multiple entry points. By separating common modules from bundles, the resulting chunked file can be loaded once initially, and stored in cache for later use. This results in pagespeed optimizations as the browser can quickly serve the shared code from cache, rather than being forced to load a larger bundle whenever a new page is visited.
+The `CommonsChunkPlugin` is an opt-in feature that creates a separate file (known as a chunk), consisting of common modules shared between multiple entry points.
+
+W> The CommonsChunkPlugin has been deprecated in webpack v4 legato. To learn how chunks are treated in the latest version, check out the [SplitChunksPlugin](/plugins/split-chunks-plugin/).
```
Comment (ooflorent) [inline src/content/plugins/commons-chunk-plugin.md]: It is not deprecated: it has been removed.

File: src/content/plugins/split-chunks-plugin.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+		}
+	}
+}
+```
+
+W> This downloads more code than neccessary.
+
+## `optimization.runtimeChunk`
+
+Setting `optimization.runtimeChunk` to `true` adds an additonal chunk to each entrypoint containing only the runtime.
```
Comment (ooflorent) [inline src/content/plugins/split-chunks-plugin.md]: Maybe you could also talk about `optimization.runtimeChunk="single"` which creates a single runtime chunk shared by all entries.

File: src/content/plugins/split-chunks-plugin.md [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+### Split Chunks: Example 2
+
+Create a `vendors` chunk, which includes all code from node_modules in the whole application.
+
+``` js
+splitChunks: {
+	cacheGroups: {
+		commons: {
+			test: /[\\/]node_modules[\\/],
```
Comment (DevSide) [inline src/content/plugins/split-chunks-plugin.md]: test: /[\\\\/]node_modules[\\\\/]**/**,

File: src/content/plugins/split-chunks-plugin.md (hunk 9 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Do we know what's the default value and behavior when this is not set?

**Comment by jeremenichelli** [comment]
Everyone here, I've updated the branch with your feedback. Feel free to take a look at it, if it is I think we should merge it already, information about this plugin should have been there from day one in my humble opinion.

**Comment by jeremenichelli** [comment]
Some things might not be covered on the first version, I'm tracking the missing stuff here: https://github.com/webpack/webpack.js.org/issues/1956

File: src/content/plugins/split-chunks-plugin.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+This configuration object represents the default behavior of the `SplitChunksPlugin`.
+
+```js
+splitChunks: {
+	chunks: "async",
+	minSize: 30000,
+	minChunks: 1,
+	maxAsyncRequests: 5,
+	maxInitialRequests: 3,
+  automaticNameDelimiter: '~',
```
Comment (lencioni) [inline src/content/plugins/split-chunks-plugin.md]: I think the indentation is wrong here. It looks like everything else is using tabs, but this line uses spaces. 
screen shot 2018-03-30 at 5 44 50 pm

Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Thanks for the catch!

File: src/content/plugins/split-chunks-plugin.md (hunk 1 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: I think that stating that they are (present tense) linked by parent-child relationship is confusing/wrong since this has completely changed on the core of the package.

File: src/content/plugins/split-chunks-plugin.md (hunk 5 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Honestly, I think that the sentence is pretty clear by itself.

**Comment by jeremenichelli** [comment]
Yes I agree, that is not clear enough. I will add that to the issue. I will
take another look during this month and do sone gardening.
-- 

*Jeremias Menichelli*


*web developer*
skype | jeremias.menichelli
<http://jeremenichelli.github.io>
github | jeremenichelli
